internal/k8s/builder/gateway.go

```diff
@@ -0,0 +1,330 @@
+		ConsulHTTPPort: orDefaultIntString(b.gwConfig.Spec.ConsulSpec.PortSpec.HTTP, defaultConsulHTTPPort),
+		ConsulGRPCPort: orDefaultIntString(b.gwConfig.Spec.ConsulSpec.PortSpec.GRPC, defaultConsulXDSPort),
+		LogLevel:       orDefault(b.gwConfig.Spec.LogLevel, defaultLogLevel),
+		GatewayHost:    "$(IP)",
+		GatewayName:    b.gateway.Name,
+		SDSHost:        b.sdsHost,
+		SDSPort:        b.sdsPort,
+	}
+	if b.requiresCA() {
+		data.ConsulCAFile = consulCALocalFile
+		data.ConsulCAData = b.consulCAData
+	}
+	if method := b.gwConfig.Spec.ConsulSpec.AuthSpec.Method; method != "" {
+		data.ACLAuthMethod = method
+	}
+	var buf bytes.Buffer
+	err := template.Must(template.New("root").Parse(strings.TrimSpace(
+		gwContainerCommandTpl))).Execute(&buf, &data)
+	if err != nil {
+		return nil
+	}
+
+	return []string{"/bin/sh", "-ec", buf.String()}
+}
+
+func (b *GatewayDeploymentBuilder) volumes() ([]corev1.Volume, []corev1.VolumeMount) {
+	volumes := []corev1.Volume{{
+		Name: "bootstrap",
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	}, {
+		Name: "certs",
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	}}
+	mounts := []corev1.VolumeMount{{
+		Name:      "bootstrap",
+		MountPath: "/bootstrap",
+	}, {
+		Name:      "certs",
+		MountPath: "/certs",
+	}}
+	if b.requiresCA() {
+		volumes = append(volumes, corev1.Volume{
+			Name: "ca",
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		})
+		mounts = append(mounts, corev1.VolumeMount{
+			Name:      "ca",
+			MountPath: consulCALocalPath,
+		})
+	}
+	return volumes, mounts
+}
+
+func (b *GatewayDeploymentBuilder) containerPorts() []corev1.ContainerPort {
+	ports := []corev1.ContainerPort{{
+		Name:          "ready",
+		Protocol:      "TCP",
+		ContainerPort: 20000,
+	}}
+	for _, listener := range b.gateway.Spec.Listeners {
+		port := corev1.ContainerPort{
+			Name:          string(listener.Name),
+			Protocol:      "TCP",
+			ContainerPort: int32(listener.Port),
+		}
+		if b.gwConfig.Spec.UseHostPorts {
+			port.HostPort = int32(listener.Port)
+		}
+		ports = append(ports, port)
+	}
+	return ports
+}
+
+func (b *GatewayDeploymentBuilder) requiresCA() bool {
+	return b.gwConfig.Spec.ConsulSpec.Scheme == "https"
+}
+
+type gwContainerCommandData struct {
+	ConsulCAFile   string
+	ConsulCAData   string
+	ConsulHTTPAddr string
+	ConsulHTTPPort string
+	ConsulGRPCPort string
+	ACLAuthMethod  string
+	LogLevel       string
+	GatewayHost    string
+	GatewayName    string
+	SDSHost        string
+	SDSPort        int
+}
+
+// gwContainerCommandTpl is the template for the command executed by
+// the exec container.
+const gwContainerCommandTpl = `
+{{- if .ConsulCAFile}}
+export CONSUL_CACERT={{ .ConsulCAFile }}
+cat <<EOF >{{ .ConsulCAFile }}
+{{ .ConsulCAData }}
+EOF
+{{- end}}
+
+/bootstrap/consul-api-gateway exec -log-json \
```

Comment:
To make sure that the signal handling actually works:

```suggestion
exec /bootstrap/consul-api-gateway exec -log-json \
```

otherwise you'll have to do some signal forwarding trickery since all of this is being invoked as a child process currently with `sh -c` -- the extra `exec` will avoid the extra child process and make sure any termination signals go directly to our binary.